app/scripts/modules/core/src/managed/Environments.tsx

```diff
@@ -1,5 +1,59 @@
-import React from 'react';
+import React, { useState, useEffect } from 'react';
 
-export default function Environments() {
-  return <div>For there shall be no greater pursuit than that towards desired state.</div>;
+import { Application, ApplicationDataSource } from '../application';
+import { IManagedApplicationSummary } from '../domain/IManagedEntity';
+import { ColumnHeader } from '../presentation/layout/ColumnHeader';
+import { ArtifactsList } from './ArtifactsList';
+import { EnvironmentsList } from './EnvironmentsList';
+
+import styles from './Environments.module.css';
+import { isEqual } from 'lodash';
+
+const CONTENT_WIDTH = 1200;
+
+export interface ISelectedArtifact {
+  name: string;
+  version: string;
+}
+
+interface IEnvironmentsProps {
+  app: Application;
+}
+
+export function Environments(props: IEnvironmentsProps) {
+  const { app } = props;
+  const dataSource: ApplicationDataSource<IManagedApplicationSummary<
```

Comment:
This caused my brain's syntax scanner to reboot:
```ts
    'resources' | 'artifacts' | 'environments'
  >> = app.getDataSource('environments');
```

This might read a lot nicer if you extract a local type
```ts
type ISummary = IManagedApplicationSummary<'resources' | 'artifacts' | 'environments'>;
```

or even just the union:
```ts
type IArtifactTypes = 'resources' | 'artifacts' | 'environments'
```

Reply:
mine too.

I added the following to the types:
```
export type IManagedApplicationEnvironmentSummary = IManagedApplicationSummary<
  'resources' | 'artifacts' | 'environments'
>;
```